File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    
    void dfs(int i,int j,vector<vector<char>> &grid,int m,int n)
    {
        if(i<0 || i>=n ||j<0 || j>=m || grid[i][j]=='0')
            return;
        grid[i][j]='0';
        dfs(i+1,j,grid,m,n);
        dfs(i-1,j,grid,m,n);
        dfs(i,j+1,grid,m,n);
        dfs(i,j-1,grid,m,n);
    }
    
    
    int numIslands(vector<vector<char>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        int ans=0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]=='1')
                {
                    ans++;
                    dfs(i,j,grid,m,n);
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    private void DFS(int i,int j,char[][] grid)
    {
       if(i<0 || j>=grid[0].length || j<0 ||i>=grid.length || grid[i][j]=='0')return;
        grid[i][j]='0';//marking cell as visited

        // explore four direction
        DFS(i+1,j,grid);
        DFS(i-1,j,grid);
        DFS(i,j+1,grid);
        DFS(i,j-1,grid);

    }
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int ans=0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]=='1')
                {
                   
                    DFS(i,j,grid);
                    ans++;
                }
            }
        }
        return ans;

    }
}